packages/loader/container-loader/src/connectionHandler.ts

```diff
@@ -0,0 +1,889 @@
+            return {
+                // Report how many ops this client sent in last disconnected session
+                sentOps: this.clientSequenceNumber,
+            };
+        }
+    }
+
+    public shouldJoinWrite(): boolean {
+        // We don't have to wait for ack for topmost NoOps. So subtract those.
+        return this.clientSequenceNumberObserved < (this.clientSequenceNumber - this.trailingNoopCount);
+    }
+
+    /**
+     * Tells if container is in read-only mode.
+     * Data stores should listen for "readonly" notifications and disallow user
+     * making changes to data stores.
+     * Readonly state can be because of no storage write permission,
+     * or due to host forcing readonly mode for container.
+     * It is undefined if we have not yet established websocket connection
+     * and do not know if user has write access to a file.
+     */
+    private get readonly() {
+        if (this._forceReadonly) {
+            return true;
+        }
+        return this._readonlyPermissions;
+    }
+
+    public get readOnlyInfo(): ReadOnlyInfo {
+        const storageOnly = this.connection !== undefined && this.connection instanceof NoDeltaStream;
+        if (storageOnly || this._forceReadonly || this._readonlyPermissions === true) {
+            return {
+                readonly: true,
+                forced: this._forceReadonly,
+                permissions: this._readonlyPermissions,
+                storageOnly,
+            };
+        }
+
+        return { readonly: this._readonlyPermissions };
+    }
+
+    private static detailsFromConnection(connection: IDocumentDeltaConnection): IConnectionDetails {
+        return {
+            claims: connection.claims,
+            clientId: connection.clientId,
+            existing: connection.existing,
+            checkpointSequenceNumber: connection.checkpointSequenceNumber,
+            get initialClients() { return connection.initialClients; },
+            maxMessageSize: connection.serviceConfiguration.maxMessageSize,
+            mode: connection.mode,
+            serviceConfiguration: connection.serviceConfiguration,
+            version: connection.version,
+        };
+    }
+
+    constructor(
+        private readonly serviceProvider: () => IDocumentService | undefined,
+        private client: IClient,
+        reconnectAllowed: boolean,
+        private readonly logger: ITelemetryLogger,
+        private readonly props: IConnectionManagereFactoryArgs,
+    ) {
+        this.clientDetails = this.client.details;
+        this.defaultReconnectionMode = this.client.mode;
+        this._reconnectMode = reconnectAllowed ? ReconnectMode.Enabled : ReconnectMode.Never;
+
+        // Outbound message queue. The outbound queue is represented as a queue of an array of ops. Ops contained
+        // within an array *must* fit within the maxMessageSize and are guaranteed to be ordered sequentially.
+        this._outbound = new DeltaQueue<IDocumentMessage[]>(
```

Comment:
Maybe the delta manager should create and own the outbound queue too for consistency? The queue can the call into `connectionManager.submitMessage`. Is there a reason why this is here?

Reply:
op pausing / resuming is controlled by this class, as well as actual sending (sendMessages() method). Plus, DM does not use it at all with this change (it re-exports it, but I think it's wrong - it's exposing too many guts and likely needs to be looked more). So, at the moment, with where the line is drawn in this PR, it makes sense to keep it here.